palantir-java-format/src/main/java/com/palantir/javaformat/java/JavaInputAstVisitor.java

```diff
@@ -1144,18 +1144,24 @@ public Void visitBinary(BinaryTree node, Void unused) {
         List<ExpressionTree> operands = new ArrayList<>();
         List<String> operators = new ArrayList<>();
         walkInfix(precedence(node), node, operands, operators);
-        FillMode fillMode = hasOnlyShortItems(operands) ? INDEPENDENT : UNIFIED;
+        boolean isStringConcat = isStringConcat(node);
+        FillMode fillMode = hasOnlyShortItems(operands) || isStringConcat ? INDEPENDENT : UNIFIED;
         builder.open(
                 plusFour,
                 BreakBehaviours.breakThisLevel(),
                 LastLevelBreakability.ACCEPT_INLINE_CHAIN_IF_SIMPLE_OTHERWISE_CHECK_INNER);
         scan(operands.get(0), null);
+        FillMode nextFillMode = builder.lastTokenFollowedByNewline() ? UNIFIED : fillMode;
         int operatorsN = operators.size();
         for (int i = 0; i < operatorsN; i++) {
-            builder.breakOp(fillMode, " ", ZERO);
+            builder.breakOp(nextFillMode, " ", ZERO);
             builder.op(operators.get(i));
+            boolean shouldEnforceNewline = builder.lastTokenFollowedByNewline();
             builder.space();
             scan(operands.get(i + 1), null);
+            shouldEnforceNewline |= builder.lastTokenFollowedByNewline();
```

Comment:
it seems like `|=` is a non-short circuiting operator... is that intentional? https://stackoverflow.com/a/2486574/1941552

Reply:
Perhaps a little comment here explaining we expect the builder to be mutated between calls

Reply:
I guess I didn't mind, but sure, can save a bit of work.